### PR TITLE
Spin: implement carousel for performance improvement

### PR DESCRIPTION
**Pull request by TomaszLukawskiSam**
[Issue] https://github.com/Samsung/TAU/issues/1139
[Problem] Scrolling performance should be better.
 In comparison to native wheel is rotating very slowly.
[Solution]
 - implement carousel
 - change drag behaviour
 - add momentum
 - configure widget for 2 items eg. AM/PM
 - added examples of Spin widget in UIComponents:
   - two item picker example
   - no loop picker example

[Known issues]
 - (sometimes) issue with stop scroll on drag - lack of "dragend" event https://github.com/Samsung/TAU/issues/1168

Signed-off-by: Tomasz Lukawski <t.lukawski@samsung.com>

### CODE REVIEW

**Comment by lslachciak** [comment]
Hello, I tested a little bit this PR with ScheduleApp
Performance is much better than before. Really good progress!
This what should be fixed is a wrong position of selected items in Date&Time picker:
image

File: examples/wearable/UIComponents/contents/user-inputs-elements/pickers/default-picker.html [hunk 1]
```diff
@@ -7,10 +7,10 @@
 
 	<link rel="stylesheet"
 			href="../../../lib/tau/wearable/theme/default/tau.css">
-	<link rel="stylesheet" media="all and (-tizen-geometric-shape: circle)"
+	<link rel="stylesheet"
```
Comment (DohyungLim) [inline examples/wearable/UIComponents/contents/user-inputs-elements/pickers/default-picker.html:10]: Was it intended to change the line?

File: examples/wearable/UIComponents/contents/user-inputs-elements/pickers/custom-picker.html [hunk 2]
```diff
@@ -7,10 +7,12 @@
 
 	<link rel="stylesheet"
 			href="../../../lib/tau/wearable/theme/default/tau.css">
-	<link rel="stylesheet" media="all and (-tizen-geometric-shape: circle)"
+	<link rel="stylesheet"
+			media="all and (-tizen-geometric-shape: circle)"
```
Comment (DohyungLim) [inline examples/wearable/UIComponents/contents/user-inputs-elements/pickers/custom-picker.html]: just curious. was it intended to change the line?

File: examples/wearable/UIComponents/contents/user-inputs-elements/pickers/custom-picker.html [hunk 3]
```diff
@@ -7,10 +7,12 @@
...
 			href="../../../lib/tau/wearable/theme/default/tau.circle.min.css">
 	<link rel="stylesheet" href="../../../css/style.css">
-	<link rel="stylesheet" media="all and (-tizen-geometric-shape: circle)"
+	<link rel="stylesheet"
+			media="all and (-tizen-geometric-shape: circle)"
```
Comment (DohyungLim) [inline examples/wearable/UIComponents/contents/user-inputs-elements/pickers/custom-picker.html]: ditto.

File: examples/wearable/UIComponents/contents/user-inputs-elements/pickers/default-picker.html [hunk 4]
```diff
@@ -7,10 +7,12 @@
...
+			media="all and (-tizen-geometric-shape: circle)"
 			href="../../../lib/tau/wearable/theme/default/tau.circle.min.css">
 	<link rel="stylesheet" href="../../../css/style.css">
-	<link rel="stylesheet" media="all and (-tizen-geometric-shape: circle)"
+	<link rel="stylesheet"
+			media="all and (-tizen-geometric-shape: circle)"
```
Comment (DohyungLim) [inline examples/wearable/UIComponents/contents/user-inputs-elements/pickers/default-picker.html]: ditto.

File: examples/wearable/UIComponents/contents/user-inputs-elements/pickers/two-items-picker.js [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+		descriptions = document.getElementById("descriptions"),
+		content = page.querySelector(".ui-content"),
+		element = page.querySelector(".ui-spin"),
+		spin = null,
+		stage = 0;
+
+	function onClickChange() {
+		switch (stage) {
+			case 0 :
+				descriptions.innerHTML = ".value(9)";
```
Comment (DohyungLim) [inline examples/wearable/UIComponents/contents/user-inputs-elements/pickers/two-items-picker.js]: what is .value(9) ?

File: examples/wearable/UIComponents/contents/user-inputs-elements/pickers/two-items-picker.js [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+				document.querySelector(".ui-footer .ui-btn")
+					.removeEventListener("click", onClick);
+				content.removeEventListener("vclick", onClick);
+
+				spin.destroy();
+			});
+		});
+
+		document.querySelector(".ui-footer .ui-btn")
+			.addEventListener("click", onClickChange);
```
Comment (DohyungLim) [inline examples/wearable/UIComponents/contents/user-inputs-elements/pickers/two-items-picker.js:100]: line check please.

File: src/css/profile/wearable/changeable/common/spin.less [hunk 7]
```diff
@@ -1,39 +1,48 @@
...
-        top: 0;
-        width: 100%;
-        height: 30 * @unit_base;
-        display: flex;
-        flex-direction: column;
-        justify-content: center;
-        color: rgba(77, 207, 255, 1);
+	position: relative;
+	display: inline-block;
+	height: 60 * @unit_base;
```
Comment (DohyungLim) [inline src/css/profile/wearable/changeable/common/spin.less]: use @px_base;

File: src/css/profile/wearable/changeable/common/spin.less [hunk 8]
```diff
@@ -1,39 +1,48 @@
...
+	background-color: transparent;
+	overflow: hidden;
+	font-size: 22 * @unit_base;
```
Comment (DohyungLim) [inline src/css/profile/wearable/changeable/common/spin.less]: use @px_base;

File: src/css/profile/wearable/changeable/common/spin.less [hunk 9]
```diff
@@ -1,39 +1,48 @@
...
+	padding: 0;
+	box-sizing: content-box;
+	width: 80 * @unit_base;
```
Comment (DohyungLim) [inline src/css/profile/wearable/changeable/common/spin.less]: use @px_base;

File: src/css/profile/wearable/changeable/common/spin.less [hunk 10]
```diff
@@ -1,39 +1,48 @@
...
+	&-item {
+		background-color: transparent;
+		width: 100%;
+		height: 30 * @unit_base;
```
Comment (DohyungLim) [inline src/css/profile/wearable/changeable/common/spin.less]: use @px_base;

File: src/css/profile/wearable/changeable/common/spin.less [hunk 11]
```diff
@@ -1,39 +1,48 @@
...
-        .ui-spin-item {
-            transition: 300ms opacity linear;
-        }
-    }
-    &-placeholder {
-        opacity: 0;
-        pointer-events: none;
-    }
+		&-selected {
+			font-size: 30 * @unit_base;
```
Comment (DohyungLim) [inline src/css/profile/wearable/changeable/common/spin.less]: use @sp_size;

File: src/css/profile/wearable/changeable/common/spin.less [hunk 12]
```diff
@@ -1,39 +1,48 @@
...
+			height: 30 * @unit_base;
```
Comment (DohyungLim) [inline src/css/profile/wearable/changeable/common/spin.less]: use @px_base;

**Review by DohyungLim — COMMENTED** [review]
12 comment.

File: examples/wearable/UIComponents/contents/user-inputs-elements/pickers/custom-picker.html (hunk 2 above)
Comment (TomaszLukawskiSam) [reply examples/wearable/UIComponents/contents/user-inputs-elements/pickers/custom-picker.html]: It was just copy/paste effect :) . It was not intended. I am trying to prepare this PR to review now.

File: examples/wearable/UIComponents/contents/user-inputs-elements/pickers/two-items-picker.js (hunk 6 above)
Comment (TomaszLukawskiSam) [reply examples/wearable/UIComponents/contents/user-inputs-elements/pickers/two-items-picker.js:100]: The line seems correct. I don't find issue.

File: src/js/core/util/animation/animation.js [hunk 13]
```diff
@@ -1,4 +1,4 @@
-/* global requestAnimationFrame, define, ns */
+/* global define, ns */
```
Comment (TomaszLukawskiSam) [inline src/js/core/util/animation/animation.js:1]: In this file have been fixed indentations. You have to change setting in github to better view of changes.

obraz

**Comment by TomaszLukawskiSam** [comment]
> Hello, I tested a little bit this PR with ScheduleApp
> Performance is much better than before. Really good progress!
> This what should be fixed is a wrong position of selected items in Date&Time picker:
> image

This issue is fixed.

File: examples/wearable/UIComponents/contents/user-inputs-elements/pickers/custom-picker.html (hunk 3 above)
Comment (TomaszLukawskiSam) [reply examples/wearable/UIComponents/contents/user-inputs-elements/pickers/custom-picker.html]: Fixed

File: examples/wearable/UIComponents/contents/user-inputs-elements/pickers/default-picker.html (hunk 4 above)
Comment (TomaszLukawskiSam) [reply examples/wearable/UIComponents/contents/user-inputs-elements/pickers/default-picker.html]: Fixed.

File: examples/wearable/UIComponents/contents/user-inputs-elements/pickers/two-items-picker.js (hunk 5 above)
Comment (TomaszLukawskiSam) [reply examples/wearable/UIComponents/contents/user-inputs-elements/pickers/two-items-picker.js]: Fixed. This is description for action from test scenario.

File: src/css/profile/wearable/changeable/common/spin.less (hunk 8 above)
Comment (TomaszLukawskiSam) [reply src/css/profile/wearable/changeable/common/spin.less]: Done.

File: src/css/profile/wearable/changeable/common/spin.less (hunk 9 above)
Comment (TomaszLukawskiSam) [reply src/css/profile/wearable/changeable/common/spin.less]: Done.

File: src/css/profile/wearable/changeable/common/spin.less (hunk 10 above)
Comment (TomaszLukawskiSam) [reply src/css/profile/wearable/changeable/common/spin.less]: Done.

File: src/css/profile/wearable/changeable/common/spin.less (hunk 11 above)
Comment (TomaszLukawskiSam) [reply src/css/profile/wearable/changeable/common/spin.less]: I cannot use `@sp_base` in wearable. I used `@px_base`

File: src/css/profile/wearable/changeable/common/spin.less (hunk 12 above)
Comment (TomaszLukawskiSam) [reply src/css/profile/wearable/changeable/common/spin.less]: Done.

File: examples/wearable/UIComponents/contents/user-inputs-elements/pickers/no-loop-picker.js [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+				/*
+				Event "spinchange" is not triggering when value is changing by .value() method.
+				eg. "rotarydetent" does not trigger "spinchange" because of value
+				    has changed by .value()
+				 */
+				descriptions.innerHTML = "on spinchange: " + spin.value();
+			});
+
+			element.addEventListener("spinstep", function () {
+				//console.log("spinstep");
```
Comment (lslachciak) [inline examples/wearable/UIComponents/contents/user-inputs-elements/pickers/no-loop-picker.js]: do we need this? or you put this just to show possibility?

File: examples/wearable/UIComponents/contents/user-inputs-elements/pickers/two-items-picker.js [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+				/*
+				Event "spinchange" is not triggering when value is changing by .value() method.
+				eg. "rotarydetent" does not trigger "spinchange" because of value
+				    has changed by .value()
+				 */
+				descriptions.innerHTML = "on spinchange: " + spin.value();
+			});
+
+			element.addEventListener("spinstep", function () {
+				//console.log("spinstep");
```
Comment (lslachciak) [inline examples/wearable/UIComponents/contents/user-inputs-elements/pickers/two-items-picker.js]: ditto.

**Review by lslachciak — APPROVED** [review]
Works fine, when working on this issue I found issue in wearable
https://github.com/Samsung/TAU/issues/1172
but I checked that this is not regression

Just minor comments/question

File: examples/wearable/UIComponents/contents/user-inputs-elements/pickers/no-loop-picker.js (hunk 14 above)
Comment (TomaszLukawskiSam) [reply examples/wearable/UIComponents/contents/user-inputs-elements/pickers/no-loop-picker.js]: removed.

File: examples/wearable/UIComponents/contents/user-inputs-elements/pickers/two-items-picker.js (hunk 15 above)
Comment (TomaszLukawskiSam) [reply examples/wearable/UIComponents/contents/user-inputs-elements/pickers/two-items-picker.js]: removed.

**Review by lslachciak — APPROVED** [review]
Rechecked latest version.
Performance seems to slightly better on wearable too.

This what we can consider more in future is quick multi swipe.
In case of native it continues spinning. In case of web it stops spinner for a moment. Can be considered as a separate issue

**Pull request merged**